share/spack/spack-completion.bash

```diff
@@ -0,0 +1,663 @@
+    then
+        compgen -W "-h --help" -- "$cur"
+    else
+        compgen -W "add create list remove" -- "$cur"
+    fi
+}
+
+function _spack_repo_add {
+    # `spack repo add` has no valid options
+    :
+}
+
+function _spack_repo_create {
+    # `spack repo create` has no valid options
+    :
+}
+
+function _spack_repo_list {
+    # `spack repo list` has no valid options
+    :
+}
+
+function _spack_repo_remove {
+    compgen -W "$(_repos)" -- "$cur"
+}
+
+function _spack_restage {
+    if $list_options
+    then
+        compgen -W "-h --help" -- "$cur"
+    else
+        compgen -W "$(_all_packages)" -- "$cur"
+    fi
+}
+
+function _spack_spec {
+    if $list_options
+    then
+        compgen -W "-h --help -i --ids" -- "$cur"
+    else
+        compgen -W "$(_all_packages)" -- "$cur"
+    fi
+}
+
+function _spack_stage {
+    if $list_options
+    then
+        compgen -W "-h --help -n --no-checksum" -- "$cur"
+    else
+        compgen -W "$(_all_packages)" -- "$cur"
+    fi
+}
+
+function _spack_test {
+    # TODO: What does this subcommand even do?
+    compgen -W "-h --help -l --list --createXmlOutput --xmlOutputDir
+                -v --verbose" -- "$cur"
+}
+
+function _spack_test-install {
+    if $list_options
+    then
+        compgen -W "-h --help -j --jobs -n --no-checksum -o --output" -- "$cur"
+    else
+        compgen -W "$(_all_packages)" -- "$cur"
+    fi
+}
+
+function _spack_uninstall {
+    if $list_options
+    then
+        compgen -W "-h --help -f --force -a --all" -- "$cur"
+    else
+        compgen -W "$(_installed_packages)" -- "$cur"
+    fi
+}
+
+function _spack_unload {
+    if $list_options
+    then
+        compgen -W "-h --help" -- "$cur"
+    else
+        compgen -W "$(_installed_packages)"
+    fi
+}
+
+function _spack_unuse {
+    if $list_options
+    then
+        compgen -W "-h --help" -- "$cur"
+    else
+        compgen -W "$(_installed_packages)"
+    fi
+}
+
+function _spack_url-parse {
+    compgen -W "-h --help -s --spider" -- "$cur"
+}
+
+function _spack_urls {
+    compgen -W "-h --help -c --color -e --extrapolation" -- "$cur"
+}
+
+function _spack_use {
+    if $list_options
+    then
+        compgen -W "-h --help" -- "$cur"
+    else
+        compgen -W "$(_installed_packages)" -- "$cur"
+    fi
+}
+
+function _spack_versions {
+    if $list_options
+    then
+        compgen -W "-h --help" -- "$cur"
+    else
+        compgen -W "$(_all_packages)" -- "$cur"
+    fi
+}
+
+function _subcommands {
+    spack help | grep "^    [a-z]" | awk '{print $1}'
+}
+
+function _all_packages {
+    spack list
+}
+
+function _installed_packages {
+    # Perl one-liner used to strip out color codes
+    spack find | grep -v "^--" | perl -pe 's/\e\[?.*?[\@-~]//g'
+}
+
+function _installed_compilers {
+    spack compilers | egrep -v "^(-|=)"
+}
+
+function _mirrors {
+    spack mirror list | awk '{print $1}'
+}
+
+function _repos {
+    spack repo list | awk '{print $1}'
+}
+
+function test_vars {
```

Comment:
`test_vars` and `pretty_print` are just testing functions. I will remove them once we've decided on a final solution and I no longer need to test anything.